src/ray/object_manager/test/object_manager_stress_test.cc

```diff
@@ -0,0 +1,395 @@
+  std::shared_ptr<gcs::AsyncGcsClient> gcs_client_1;
+  std::shared_ptr<gcs::AsyncGcsClient> gcs_client_2;
+  std::unique_ptr<MockServer> server1;
+  std::unique_ptr<MockServer> server2;
+
+  plasma::PlasmaClient client1;
+  plasma::PlasmaClient client2;
+  std::vector<ObjectID> v1;
+  std::vector<ObjectID> v2;
+};
+
+class StressTestObjectManager : public TestObjectManagerBase {
+ public:
+  enum TransferPattern {
+    PUSH_A_B,
+    PUSH_B_A,
+    BIDIRECTIONAL_PUSH,
+    PULL_A_B,
+    PULL_B_A,
+    BIDIRECTIONAL_PULL,
+  };
+
+  int async_loop_index = -1;
+  uint num_expected_objects;
+
+  std::vector<TransferPattern> async_loop_patterns = {
+      PUSH_A_B, PUSH_B_A, BIDIRECTIONAL_PUSH, PULL_A_B, PULL_B_A, BIDIRECTIONAL_PULL};
+
+  int num_connected_clients = 0;
+
+  ClientID client_id_1;
+  ClientID client_id_2;
+
+  void WaitConnections() {
+    client_id_1 = gcs_client_1->client_table().GetLocalClientId();
+    client_id_2 = gcs_client_2->client_table().GetLocalClientId();
+    gcs_client_1->client_table().RegisterClientAddedCallback(
+        [this](gcs::AsyncGcsClient *client, const ClientID &id,
+               const ClientTableDataT &data) {
+          ClientID parsed_id = ClientID::from_binary(data.client_id);
+          if (parsed_id == client_id_1 || parsed_id == client_id_2) {
+            num_connected_clients += 1;
+          }
+          if (num_connected_clients == 2) {
+            StartTests();
+          }
+        });
+  }
+
+  void StartTests() {
+    TestConnections();
+    AddTransferTestHandlers();
+    TransferTestNext();
+  }
+
+  void AddTransferTestHandlers() {
+    ray::Status status = ray::Status::OK();
+    status =
+        server1->object_manager_.SubscribeObjAdded([this](const ObjectID &object_id) {
+          object_added_handler_1(object_id);
+          if (v1.size() == num_expected_objects && v1.size() == v2.size()) {
+            TransferTestComplete();
+          }
+        });
+    RAY_CHECK_OK(status);
+    status =
+        server2->object_manager_.SubscribeObjAdded([this](const ObjectID &object_id) {
+          object_added_handler_2(object_id);
+          if (v2.size() == num_expected_objects && v1.size() == v2.size()) {
+            TransferTestComplete();
+          }
+        });
+    RAY_CHECK_OK(status);
+  }
+
+  void TransferTestNext() {
+    async_loop_index += 1;
+    if ((uint)async_loop_index < async_loop_patterns.size()) {
+      TransferPattern pattern = async_loop_patterns[async_loop_index];
+      TransferTestExecute(1000, 100, pattern);
+    } else {
+      main_service.stop();
+    }
+  }
+
+  plasma::ObjectBuffer GetObject(plasma::PlasmaClient &client, ObjectID object_id) {
+    plasma::ObjectBuffer object_buffer;
+    plasma::ObjectID plasma_id = object_id.to_plasma_id();
+    ARROW_CHECK_OK(client.Get(&plasma_id, 1, 0, &object_buffer));
+    return object_buffer;
+  }
+
+  static unsigned char *GetDigest(plasma::PlasmaClient &client, ObjectID object_id) {
```

Comment:
const refs on ObjectIDs in GetObject, GetDigest, CompareObjects, CompareHashes -- avoid invoking a copy ctor whenever possible.